watchpoints.c

```diff
@@ -133,79 +126,71 @@ static long watchpoints_ioctl(struct file *file, unsigned int cmd, long
 	       data.pid, data.data_ptr, data.data_size);
 
 	switch(cmd) {
-		struct task_struct *tsk;
-		
-		case ADD_BREAKPOINT:
-			// Initialize breakpoint
-			hw_breakpoint_init(&attr);
-			attr.bp_addr = data.data_ptr;
-			attr.bp_len = HW_BREAKPOINT_LEN_4;
-			attr.bp_type = HW_BREAKPOINT_W;
-
-			tsk = pid_task(find_vpid(data.pid), PIDTYPE_PID);
-
-			perf_watchpoint = register_user_hw_breakpoint(&attr,
-					watchpoint_handler, NULL, tsk);
+	case ADD_BREAKPOINT:
+		// Initialize breakpoint
+		struct perf_event_attr attr;
+		hw_breakpoint_init(&attr);
+		attr.bp_addr = data.data_ptr;
+		attr.bp_len = HW_BREAKPOINT_LEN_4;
+		attr.bp_type = HW_BREAKPOINT_W;
+
+		struct task_struct *tsk = pid_task(find_vpid(data.pid), PIDTYPE_PID);
+
+		struct perf_event *perf_watchpoint = 
+			register_user_hw_breakpoint(&attr, watchpoint_handler, NULL, tsk);
 			
-			if (IS_ERR(perf_watchpoint)) {
-				printk(KERN_DEBUG "Could not set watchpoint");
-				return perf_watchpoint;
-			}
+		if (IS_ERR(perf_watchpoint)) {
+			printk(KERN_DEBUG "Could not set watchpoint");
+			return perf_watchpoint;
+		}
 
-			for(i = 0; i < WATCHPOINTS_MAX; i++) {
-				if(! watchpoints[i]) {
-					watchpoints[i] = perf_watchpoint;
-					watchpoint_data_size[i] = data.data_size;
-					break;
-				}
-				else if(watchpoints[i]->state ==
-					PERF_EVENT_STATE_OFF) {
-					printk(KERN_DEBUG 
+		for(int i = 0; i < WATCHPOINTS_MAX; i++) {
+			if(watchpoints[i] && 
+			   watchpoints[i]->state == PERF_EVENT_STATE_OFF) {
+				printk(KERN_DEBUG 
 					"Removing watchpoint at %i. Not used anymore\n", i);
-					
-					unregister_hw_breakpoint(watchpoints[i]);
-					watchpoints[i] = perf_watchpoint;
-					watchpoint_data_size[i] = data.data_size;
-					break;
-				}
+				unregister_hw_breakpoint(watchpoints[i]);
+				watchpoints[i] = NULL;
 			}
-			break;
-
-		case REMOVE_BREAKPOINT:
-			for(i = 0; i < WATCHPOINTS_MAX; i++) {
-				if(watchpoints[i] &&
-				    watchpoints[i]->attr.bp_addr ==
-				    data.data_ptr &&
-				    watchpoints[i]->ctx->task->pid == data.pid) {
-					unregister_hw_breakpoint(watchpoints[i]);
-				}
+			if(!watchpoints[i]) {
+				watchpoints[i] = perf_watchpoint;
+				watchpoint_data_size[i] = data.data_size;
+				break;
 			}
+		}
+		break;
 
-		default:
-			return -EINVAL;
-
+	case REMOVE_BREAKPOINT:
+		for(int i = 0; i < WATCHPOINTS_MAX; i++) {
+			if(watchpoints[i] &&
+			   watchpoints[i]->attr.bp_addr == data.data_ptr &&
+			   watchpoints[i]->ctx->task->pid == data.pid) {
+				unregister_hw_breakpoint(watchpoints[i]);
+			}
+		}
```

Comment:
The lack of a break here seems like a bug; do you always want to return EINVAL when removing a breakpoint?